### PR TITLE
Bring back product metatags

### LINKED ISSUE #566: Adds logic to bring back the og metatags #802
With the upcoming release of Yoast SEO 14.0 the filter wpseo_opengraph is deprecated. Woocommerce SEO is hooking into this filter for rendering OG metatags. We should replaces these by presenters.

For reference see:
https://github.com/Yoast/wpseo-woocommerce/blob/c0ea263345da47cb3b58c3855a5f03ee3cf5b0da/classes/woocommerce-opengraph.php#L21

https://github.com/Yoast/wpseo-woocommerce/blob/c0ea263345da47cb3b58c3855a5f03ee3cf5b0da/classes/woocommerce-opengraph.php#L72

### PR DESCRIPTION
**Pull request by igorschoester**
## TODO

* Write tests

## Context

* We should...
  1. ...still output our metatags as before
  2. ...not trigger our own deprecation notices
  3. ...show deprecation notices when the functions are still called by external parties

## Summary

This PR can be summarized in the following changelog entry:

* Adds a WooCommerce presenter to bring back the product metatags that relied on the (deprecated) `wpseo_opengraph` action.

## Relevant technical choices:

* I'm not sure about...
  * ...the location of the deprecation of the `Yoast\WP\Woocommerce\opengraph` action. I put it there because the old location (`product_enhancement`) is deprecated too.
  * ...the removal of the action in the old `product_enhancement` method.
* I chose not to update too much, if you spot some easy fixes, feel free to apply them 😉 

## Test instructions

This PR can be tested by following these steps:

* Install WooCommerce and WooCommerce Brands and set up a product with a brand, price, availability and so on.
* Compare the metatag output before this branch and after this branch, there should be:
  - product:brand
  - product:price:amount
  - product:price:currency
  - og:availability
  - product:availability
  - product:retailer_item_id
  - product:condition
* The metatag of `product:availability` when in stock is changed compared to what it was before. From `in stock` to `instock`. As this got noted as being wrong by Jono, see: https://github.com/Yoast/wpseo-woocommerce/issues/555 for future changes.
* You could manually run the actions that are now deprecated to see the notices.

Fixes #566 
Fixes https://github.com/Yoast/bugreports/issues/950

### CODE REVIEW

File: classes/woocommerce-presenter.php [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+		}
+
+		/**
+		 * Action: 'Yoast\WP\Woocommerce\opengraph' - Allow developers to add to our OpenGraph tags.
+		 *
+		 * @since 12.6.0
+		 *
+		 * @api   WC_Product $product The WooCommerce product we're outputting for.
+		 */
+		do_action_deprecated( 'Yoast\WP\Woocommerce\opengraph', $product, 'WPSEO Woo 13.0', 'WPSEO_WooCommerce_Presenter' );
```
Comment (andizer) [inline classes/woocommerce-presenter.php:33]: Why is this deprecated? This means there is no way for third parties to output extra OG-tags?

Comment (herregroen) [reply classes/woocommerce-presenter.php:33]: I discussed this with Igor and felt that the existing filters should be sufficient.

The only thing that's wrong here is the replacement. That should be the `wpseo_frontend_presenters` filter.

**Comment by hansjovis** [comment]
Acceptance test: 🚫 

When I try to view the product on the frontend, I get a white screen, nothing else 😕. 
(With `release/14.0` of `wordpress-seo` checked out).

When I deactivate all other plugins except WooCommerce, Wordpress SEO and WooCommerce SEO I get this fatal error:

```
Fatal error: Class WPSEO_WooCommerce_Presenter contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (Yoast\WP\SEO\Presenters\Abstract_Indexable_Presenter::get)
```

The `Abstract_Indexable_Presenter` has changed last Thursday. It now expects a `get` method to be implemented. I expect Herre has more context on what has changed.

File: classes/woocommerce-abstract-product-availability-presenter.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+	/**
+	 * Whether the product is in stock.
+	 *
+	 * @var bool
+	 */
+	protected $is_in_stock;
+
+	/**
+	 * WPSEO_WooCommerce_Abstract_Product_Presenter constructor.
```
Comment (andizer) [inline classes/woocommerce-abstract-product-availability-presenter.php:28]:
```suggestion
	 * WPSEO_WooCommerce_Abstract_Product_Availability_Presenter constructor.
```

File: classes/woocommerce-opengraph.php [hunk 3]
```diff
@@ -154,6 +139,8 @@ protected function get_brand_term_name( $schema_brand, $product ) {
 	 * @param WC_Product $product The WooCommerce product object.
 	 */
 	public function brand( WC_Product $product ) {
+		_deprecated_function( __METHOD__, 'WPSEO Woo 13.0', 'WPSEO_WooCommerce_Presenter::product_brand' );
```
Comment (andizer) [inline classes/woocommerce-opengraph.php:142]: Don't forget to replace this and the other ones referring to `WPSEO_WooCommerce_Presenter` with the actual presenters.

File: classes/woocommerce-product-availability-presenter.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * WooCommerce Yoast SEO plugin file.
+ *
+ * @package WPSEO/WooCommerce
+ */
+
+use WPSEO_WooCommerce_Abstract_Product_Availability_Presenter;
+
+/**
+ * Represents the product's price currency.
```
Comment (johannadevos) [inline classes/woocommerce-product-availability-presenter.php:11]:
```suggestion
 * Represents the product's availability.
```

File: classes/woocommerce-product-retailer-item-id-presenter.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * WooCommerce Yoast SEO plugin file.
+ *
+ * @package WPSEO/WooCommerce
+ */
+
+use WPSEO_WooCommerce_Abstract_Product_Presenter;
+
+/**
+ * Represents the product's retail item ID.
```
Comment (johannadevos) [inline classes/woocommerce-product-retailer-item-id-presenter.php:11]:
```suggestion
 * Represents the product's retailer item ID.
```

File: classes/woocommerce-product-availability-presenter.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	protected $tag_format = '<meta property="product:availability" content="%s" />';
+
+	/**
+	 * Gets the raw value of a presentation.
+	 *
+	 * @return string The raw value.
+	 */
+	public function get() {
+		if ( $this->is_on_backorder ) {
+			return 'available for order';
```
Comment (johannadevos) [inline classes/woocommerce-product-availability-presenter.php:29]: In `classes/woocommerce-pinterest-product-availability-presenter.php` you return `'backorder'`, here you return `'available for order'`. For consistency, should it not be the same string?

File: classes/woocommerce-seo.php [hunk 7]
```diff
@@ -1220,4 +1239,40 @@ public function upgrade() {
...
+	 *
+	 * @return bool True when the price should be shown.
+	 */
+	private function should_show_price() {
+		/**
+		 * Filter: wpseo_woocommerce_og_price - Allow developers to prevent the output of the price in the OpenGraph tags.
+		 *
+		 * @deprecated 12.5.0. Use the {@see 'Yoast\WP\Woocommerce\og_price'} filter instead.
+		 *
+		 * @api        bool unsigned Defaults to true.
```
Comment (johannadevos) [inline classes/woocommerce-seo.php:1254]:
```suggestion
		 * @api bool unsigned Defaults to true.
```

File: classes/woocommerce-seo.php [hunk 8]
```diff
@@ -1220,4 +1239,40 @@ public function upgrade() {
...
+			'Yoast WooCommerce 12.5.0',
+			'Yoast\WP\Woocommerce\og_price'
+		);
+
+		/**
+		 * Filter: Yoast\WP\Woocommerce\og_price - Allow developers to prevent the output of the price in the OpenGraph tags.
+		 *
+		 * @since 12.5.0
+		 *
+		 * @api   bool unsigned Defaults to true.
```
Comment (johannadevos) [inline classes/woocommerce-seo.php:1268]:
```suggestion
		 * @api bool unsigned Defaults to true.
```

File: classes/woocommerce-product-availability-presenter.php (hunk 6 above)
Comment (igorschoester) [reply classes/woocommerce-product-availability-presenter.php:29]: No, unfortunately, this is not in our control. It is how they implemented their metatags.

**Comment by johannadevos** [comment]
CR: ✅ , except there is one unfinished task. 
Note: I only reviewed commits `240aff2` and `a77480b` because @hansjovis already did a CR of the rest.
* [x] The `meta` surface is now called correctly.
* [x] The refactoring of the presenters looks good to me.
* [x] All the new presenters are added in `woocommerce-seo.php`. 
* [x] I pushed three commits with small fixes and improvements.
* [ ] @igorschoester will finish replacing the presenters in the deprecated functions.

**Comment by hansjovis** [comment]
Acceptance test: 🚫 

The value of the `product:availability` meta tag is different when the product is in stock.
In this branch it is `instock`, whereas it is `in stock` in `release/13.0`:
```html
<meta property="product:availability" content="instock">
```
vs
```html
<meta property="product:availability" content="in stock">
```

It works for the rest 🙂.

**Comment by hansjovis** [comment]
> The value of the product:availability meta tag is different when the product is in stock.
In this branch it is `instock`, whereas it is `in stock` in release/13.0

This works as expected, as noted by Jono (see #555 for future changes).

**Pull request merged**